jupyterhub/metrics.py

```diff
@@ -200,3 +230,69 @@ def prometheus_log_method(handler):
         handler=f'{handler.__class__.__module__}.{type(handler).__name__}',
         code=handler.get_status(),
     ).observe(handler.request.request_time())
+
+
+class PeriodicMetricsCollector(LoggingConfigurable):
+    """
+    Collect metrics to be calculated periodically
+    """
+
+    active_users_metrics_enabled = Bool(
+        True,
+        help="""
+        Enable active_users prometheus metric
+
+        Populates an `active_users` prometheus metric, with a label `period` that counts the time period
+        over which these many users were active. Defaults to 24h (24 hours), 7d (7 days) and 30d (30 days).
+        """,
+        config=True,
+    )
+
+    active_users_metrics_update_period = Integer(
+        60 * 60,
+        help="""
+        Number of seconds between updating active_users.
+
+        To avoid extra load on the database, this is only calculated periodically rather than
+        at per-minute intervals. Defaults to once an hour.
+        """,
+        config=True,
+    )
+
+    db = Any(help="SQLAlchemy db to use for performing queries")
+
+    def update(self):
```

Comment:
It seems fine that the class is designed ahead of time to be general purpose for other metrics to updated periodically, but I think the update function needs to be specific to active_users. So `update_active_users` for example?

The `start` function is called, and checks if active_users_metrics_enabled is truthy, and then a periodic callback is scheduled to run this `update` function. So that makes me think it should be `update_active_users` rather than a general `update` function.

Reply:
I'm torn! I think `update` is the right name. This is a generic class for periodically updating metrics. Active users happens to be the only kind we have right now, so that's the only checks we have, but adding more wouldn't change what the update method should be called. It might move exactly where the `enabled` checks happen, though.

Reply:
It sounds like this class is trying to be both an Abstract class (`start()` and `update()` methods) and a concrete implementation (defines `active_users_metrics_enabled` and `active_users_metrics_update_period` which will be inherited by any subclasses of `PeriodicMetricsCollector` even if it's used for some other metrics. Would it make sense to separate out the Abstract class, or is that over engineering things?

Reply:
Update can be the name, but then the active_users_metrics_update_period is a bit off.

Since its now a class designed for future metrics to be updated periodically, i think it should be natural on how to go about it. But, there can only be ome update period as its done now right - that is why either the update name or the active_users_metrics_update_period doesnt feel right.

Reply:
In other words, pretend another metrics is added to use this class. How should it be added? Maybe it was meant as a base class, and when a new class is added we need to extract active_users_metric, or maybe it was meant as a single class with one update interval, or a class that should handle multiple update intervals. A decision should be made now, or a comment indicating that whoever adds something in the future can do as they think makes sense.

Reply:
> either the update name or the active_users_metrics_update_period doesnt feel right.

👍 I agree, they should probably match. Since @yuvipanda named the config value with `active_users`, I made the method name match, so presumably adding another periodically updated metric will mean a new config for enabled/interval and a new update_$metrics method.

> It sounds like this class is trying to be both an Abstract class

@manics I don't think it's meant to be an abstract class at all, I think it's meant to be the one and only periodic metrics updater instance. So if we have other periodically updated metrics, they would be added to the update in this class, not a new class. I think the only question is whether the period should be shared across all metrics or per metric group, and naming follows from that. Naming of methods is a private implementation detail and not something I'd worry about, but naming of config is a user-facing API we need to consider because it's a pain to change.

Reply:
Yep, it was meant to be sort of a singleton, rather than an abstract class. I am extremely happy with the state it was merged in! If we add more of these metrics in the future, i think it makes sense to add individual traitlet knobs for those, and we can even coalesce periodcallbacks if needed.

Thank you so much @minrk for taking this to completion.